### PR TITLE
refactor: upgrade to PHP 8.1 with rector

### PR DESCRIPTION
**Pull request by kenjis**
**Description**
Supersedes #6923
See #6921 and #8042

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: app/Views/errors/html/error_exception.php [hunk 1]
```diff
@@ -90,7 +90,7 @@
                                         <?php
                                         $params = null;
                                         // Reflection by name is not available for closure function
-                                        if (substr($row['function'], -1) !== '}') {
+                                        if (!str_ends_with($row['function'], '}')) {
```
Comment (kenjis) [inline app/Views/errors/html/error_exception.php:93]: Why is there no space after `!`?

File: system/Cache/Handlers/PredisHandler.php [hunk 2]
```diff
@@ -86,22 +86,11 @@ public function get(string $key)
...
         }
 
-        switch ($data['__ci_type']) {
-            case 'array':
-            case 'object':
-                return unserialize($data['__ci_value']);
-
-            case 'boolean':
-            case 'integer':
-            case 'double': // Yes, 'double' is returned and NOT 'float'
```
Comment (kenjis) [inline system/Cache/Handlers/PredisHandler.php:96]: This comment is gone.

File: app/Views/errors/html/error_exception.php (hunk 1 above)
Comment (samsonasik) [reply app/Views/errors/html/error_exception.php:93]: that should be a coding standard tool for that :)

File: system/Cache/Handlers/RedisHandler.php [hunk 3]
```diff
@@ -112,22 +112,11 @@ public function get(string $key)
...
         }
 
-        switch ($data['__ci_type']) {
-            case 'array':
-            case 'object':
-                return unserialize($data['__ci_value']);
-
-            case 'boolean':
-            case 'integer':
-            case 'double': // Yes, 'double' is returned and NOT 'float'
```
Comment (kenjis) [inline system/Cache/Handlers/RedisHandler.php:122]: This comment is gone. Do we need it?

File: system/Cache/Handlers/PredisHandler.php (hunk 2 above)
Comment (samsonasik) [reply system/Cache/Handlers/PredisHandler.php:96]: comment is belong to next stmt, this gone as expected, should be manually added if going to re-add.

File: system/Cache/Handlers/RedisHandler.php (hunk 3 above)
Comment (samsonasik) [reply system/Cache/Handlers/RedisHandler.php:122]: comment is belong to next stmt, this gone as expected, should be manually added if going to re-add.

File: app/Views/errors/html/error_exception.php (hunk 1 above)
Comment (kenjis) [reply app/Views/errors/html/error_exception.php:93]: Yes, but I ran php-cs-fixer after running rector.

Comment (samsonasik) [reply app/Views/errors/html/error_exception.php:93]: I am not sure about that

File: system/Common.php [hunk 4]
```diff
@@ -384,22 +373,13 @@ function env(string $key, $default = null)
             return $default;
         }
 
-        // Handle any boolean values
```
Comment (samsonasik) [inline system/Common.php:387]: comment should be copied, this will be covered at Rector side:

- https://github.com/rectorphp/rector-src/pull/5072

**Comment by kenjis** [comment]
@samsonasik When I ran rector first, the following change did not happen.
I ran rector again, I got the change.

```diff
$ vendor/bin/rector 
 119/119 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
1 file with changes
===================

1) system/Router/AutoRouterImproved.php:410

    ---------- begin diff ----------
@@ @@
     {
         try {
             $refClass = new ReflectionClass($this->controller);
-        } catch (ReflectionException $e) {
+        } catch (ReflectionException) {
             throw PageNotFoundException::forControllerNotFound($this->controller, $this->method);
         }
    ----------- end diff -----------

Applied rules:
 * RemoveUnusedVariableInCatchRector (https://wiki.php.net/rfc/non-capturing_catches)
```

File: app/Views/errors/html/error_exception.php (hunk 1 above)
Comment (kenjis) [reply app/Views/errors/html/error_exception.php:93]: Ah, it is a HTML file, so php-cs-fixer does not fix?

**Comment by samsonasik** [comment]
@kenjis rector sometime need twice run as the code got consecutive run on single rule or just reprinted on next rule apply to specific node to avoid infinite loop

**Comment by kenjis** [comment]
@samsonasik I got it. Thanks!

File: app/Views/errors/cli/error_exception.php [hunk 5]
```diff
@@ -41,20 +41,11 @@
...
-        $args = implode(', ', array_map(static function ($value) {
-            switch (true) {
-                case is_object($value):
-                    return 'Object(' . get_class($value) . ')';
-
-                case is_array($value):
-                    return count($value) ? '[...]' : '[]';
-
-                case $value === null:
-                    return 'null'; // return the lowercased version
```
Comment (kenjis) [inline app/Views/errors/cli/error_exception.php:53]: Do we need this comment?

Comment (samsonasik) [reply app/Views/errors/cli/error_exception.php:53]: I think not, if needed, rector can't cover it automatically as moved inside match Expr

File: system/View/Cells/Cell.php [hunk 6]
```diff
@@ -118,7 +119,7 @@ final protected function view(?string $view, array $data = []): string
     /**
      * Provides capability to render on string casting.
      */
-    public function __toString()
+    public function __toString(): string
```
Comment (kenjis) [inline system/View/Cells/Cell.php:122]: No breaking change.
https://3v4l.org/qKR5s

File: app/Views/errors/cli/error_exception.php (hunk 5 above)
Comment (kenjis) [reply app/Views/errors/cli/error_exception.php:53]: No problem. 
My question is "Is it worth to keep the comment manually?".

Comment (samsonasik) [reply app/Views/errors/cli/error_exception.php:53]: I think not

File: system/View/Cells/Cell.php (hunk 6 above)
Comment (samsonasik) [reply system/View/Cells/Cell.php:122]: Yes 👍

File: system/Common.php [hunk 7]
```diff
@@ -384,22 +373,13 @@ function env(string $key, $default = null)
...
-
-            case 'empty':
-                return '';
-
-            case 'null':
-                return null;
-        }
-
-        return $value;
+        return match (strtolower($value)) {
```
Comment (samsonasik) [inline system/Common.php:376]:
```suggestion
        // Handle any boolean values
        return match (strtolower($value)) {
```

Comment (kenjis) [reply system/Common.php:377]: Done.

File: system/Cache/Handlers/PredisHandler.php (hunk 2 above)
Comment (kenjis) [reply system/Cache/Handlers/PredisHandler.php:96]: Added the comment manually.

File: system/Cache/Handlers/RedisHandler.php (hunk 3 above)
Comment (kenjis) [reply system/Cache/Handlers/RedisHandler.php:122]: Added the comment manually.

**Comment by kenjis** [comment]
Rebased and re-run rector.

File: app/Views/errors/html/error_exception.php (hunk 1 above)
Comment (MGatner) [reply app/Views/errors/html/error_exception.php]: I believe we have Views in the ignore path, right?

**Review by samsonasik — APPROVED** [review]

File: app/Views/errors/cli/error_exception.php (hunk 5 above)
Comment (MGatner) [reply app/Views/errors/cli/error_exception.php:53]: Agreed, unnecessary.

**Comment by MGatner** [comment]
I'm not going to review all the changes, but what I spot-checked looks good and pipeline is passing. 

This will inevitably be a big merge conflict before we are ready for the next minor release. I suggest you get the automated changes in place via Rector (done) then consolidate the manual revisions to a commit(s) that can be cherry-picked after rerunning Rector. 

Excited to be on 8+! Thanks for your work on this - and shoutout to the attention from @TomasVotruba 😍 and the tireless effort from the whole Rector team, including our own @samsonasik.

**Comment by kenjis** [comment]
@MGatner I don't understand what you exactly mean. 
I did atomic commits. Do you suggest to combine the commit for refactor by Rector and the commits for manual tweaks into one commit?

**Comment by MGatner** [comment]
No, sorry to be unclear! I think what you are doing is great - I was just noting that it might not be worthwhile to keep up with conflicts on this PR, but rather wait and rerun Rector right before we are ready for release.

**Comment by kenjis** [comment]
@MGatner I got your point. Thanks.

File: rector.php [hunk 8]
```diff
@@ -50,7 +57,7 @@
 return static function (RectorConfig $rectorConfig): void {
     $rectorConfig->sets([
         SetList::DEAD_CODE,
-        LevelSetList::UP_TO_PHP_74,
+        LevelSetList::UP_TO_PHP_80,
```
Comment (samsonasik) [inline rector.php:60]: it can use up to php 8.1 since php 8.0 no longer support on 4.5 branch

Comment (kenjis) [reply rector.php]: Done.

**Comment by kenjis** [comment]
@samsonasik This seems a bug in rector.
```diff
2) system/Test/FilterTestTrait.php:142

    ---------- begin diff ----------
@@ @@
             $filter = new $class();

             if (! $filter instanceof FilterInterface) {
-                throw FilterException::forIncorrectInterface(get_class($filter::class));
+                throw FilterException::forIncorrectInterface($filter::class::class);
             }
         }
    ----------- end diff -----------

Applied rules:
 * ClassOnObjectRector (https://wiki.php.net/rfc/class_name_literal_on_object)
```
https://github.com/codeigniter4/CodeIgniter4/actions/runs/6606559999/job/17942876910?pr=7964

**Comment by samsonasik** [comment]
@kenjis looking at mobile, what the expected diff?

**Comment by kenjis** [comment]
Don't mind. You probably don't need to care about it.
Either I accidentally changed the code when I did `git rebase`, or there was a bug in a past Rector that has been fixed.

The `$filter::class::class` is invalid.
But to begin with, the code `get_class($filter::class)` is invalid. 
See https://3v4l.org/Gfb7C

According to the commit log, it is `refactor: upgrade to PHP 8.0 by rector`.
But I tried now, Rector did not refactor such invalid code.

```diff
    ---------- begin diff ----------
@@ @@
             $filter = new $class();

             if (! $filter instanceof FilterInterface) {
-                throw FilterException::forIncorrectInterface(get_class($filter));
+                throw FilterException::forIncorrectInterface($filter::class);
             }
         }
    ----------- end diff -----------
```

File: utils/Rector/UnderscoreToCamelCaseVariableNameRector.php [hunk 9]
```diff
@@ -39,16 +39,12 @@ final class UnderscoreToCamelCaseVariableNameRector extends AbstractRector
...
 
-    private ReservedKeywordAnalyzer $reservedKeywordAnalyzer;
-    private PhpDocInfoFactory $phpDocInfoFactory;
     private bool $hasChanged = false;
 
     public function __construct(
-        ReservedKeywordAnalyzer $reservedKeywordAnalyzer,
-        PhpDocInfoFactory $phpDocInfoFactory
+        private readonly ReservedKeywordAnalyzer $reservedKeywordAnalyzer,
+        private PhpDocInfoFactory $phpDocInfoFactory
```
Comment (samsonasik) [inline utils/Rector/UnderscoreToCamelCaseVariableNameRector.php:46]: readonly is ok here:

```suggestion
        private readonly PhpDocInfoFactory $phpDocInfoFactory
```

Comment (kenjis) [reply utils/Rector/UnderscoreToCamelCaseVariableNameRector.php]: Thanks! Done.

**Comment by kenjis** [comment]
Due to the difficulty in resolving the conflict, I close this PR.

**Pull request closed without merging**